kotlinx-coroutines-core/common/src/channels/Produce.kt

```diff
@@ -25,6 +25,32 @@ public interface ProducerScope<in E> : CoroutineScope, SendChannel<E> {
     val channel: SendChannel<E>
 }
 
+/**
+ * Suspends the current coroutine until the channel is either [closed][SendChannel.close] or [cancelled][ReceiveChannel.cancel]
+ * and invokes the given [block] before resuming the coroutine.
+ *
+ * Example of usage:
+ * ```
+ * val callbackEventsStream = produce {
+ *     val disposable = registerChannelInCallback(channel)
+ *     await { disposable.dispose() }
+ * }
+ * ```
+ */
+@ExperimentalCoroutinesApi
+public suspend fun <T> ProducerScope<T>.await(block: () -> Unit = {}) {
+    check(kotlin.coroutines.coroutineContext[Job] === this) { "await() can be invoke only from the producer context" }
+    suspendCancellableCoroutine<Unit> { cont ->
+        invokeOnClose {
+            try {
+                block()
```

Comment:
There is no guarantee for exectuion context of the `block` which might lead to nasty surprives when trying to "undo" something in UI on cancellation of UI-confined producer. I'd suggest the following reimplementation:

```
    try {
        suspendCancellableCoroutine<Unit> { cont ->
            invokeOnClose {
                cont.resume(Unit)
            }
        }
    } finally {
        block()
    }
```